vcenter_operator/configurator.py

```diff
@@ -316,21 +322,37 @@ def setup_os_session(self):
         )
         self.os_session = Session(auth=auth)
 
-    def _poll_nova(self):
-        if not self.os_session:
-            return
+    def _poll_nova_cells(self):
+        namespace_nova = self.global_options['namespace']
 
+        label_selector = 'system=openstack,component=nova,type=nova-cell'
         try:
-            endpoint_filter = {'service_type': 'compute', 'interface': 'public'}
-            resp = self.os_session.get('/os-cells', endpoint_filter=endpoint_filter)
-            for cell in resp.json().get('cellsv2', []):
-                self.global_options['cells'][cell['name']] = cell
-        except (HttpError, ConnectionError) as e:
-            LOG.error(f"Failed to get cells: {e}")
+            secrets = client.CoreV1Api().list_namespaced_secret(
+                namespace=namespace_nova,
+                label_selector=label_selector)
+        except client.ApiException as e:
+            LOG.error(f"Failed to retrieve secrets with labels {label_selector} from ns {namespace_nova}: {e}")
+            return False
+
+        if not secrets.items:
+            return
+
+        for secret in secrets.items:
+            if secret.type != 'Opaque':
+                LOG.error(f"Unknown secret type {secret.type} for {namespace_nova}/{secret.metadata.name}. "
+                          "Can only work with Opaque.")
+                return False
```

Comment:
Are we sure there are no other secrets matching our labels? I guess they're pretty specific, and it's better to fail early, like you said in the commit message. But I still wonder how the code or the labels might change in the future.

There really is no way to check whether we have all the cells we want (and just ignore other secrets that somehow made the list) since this is ground truth, isn't it? It just feels brittle, but then again, so does much of our infrastructure. 😅

Reply:
Yes, there is no way to check we have all the cells we could have, because the source-of-truth is basically these secrets. If Nova is defining broken information, then we can't do anything but act on it.

I haven't seen any other secret of that type in helm-charts and I would doubt there is any other. Even thought someone might define the same labels in the future, I think that someone should notice that we already have secrets with the same. Therefore, I don't feel too back about these labels.
```
$ git grep 'type: nova-cell'
openstack/nova/templates/_cell_secret.yaml.tpl:    type: nova-cell
```